File: chat_example/tcp/java/ServerTCP.java
/**
 * ServerTCP.java — Java TCP Chat Server
 * Computer Networks Lab
 *
 * Listens for TCP client connections on port 5556.
 * Every message received from one client is broadcast to all other clients.
 *
 * Compile:  javac ServerTCP.java
 * Run:      java ServerTCP
 */

import java.io.*;
import java.net.*;
import java.util.*;
import java.util.concurrent.*;

public class ServerTCP {

    static final int PORT = 5556;

    // Thread-safe map: connection id → client handler
    static final ConcurrentHashMap<Integer, ClientHandler> clients = new ConcurrentHashMap<>();
    static int nextId = 0;

    static synchronized int allocId() {
        return ++nextId;
    }

    // ── broadcast to every client except the sender ───────────────────────────
    static void broadcast(String message, int excludeId) {
        System.out.println("[server] " + message);
        byte[] data = (message + "\n").getBytes();
        for (Map.Entry<Integer, ClientHandler> entry : clients.entrySet()) {
            if (entry.getKey() == excludeId) continue;
            try {
                entry.getValue().out.write(data);
                entry.getValue().out.flush();
            } catch (IOException e) {
                System.out.println("[server] could not reach #" + entry.getKey());
            }
        }
    }

    // ── entry point ───────────────────────────────────────────────────────────
    public static void main(String[] args) throws IOException {
        ServerSocket serverSocket = new ServerSocket(PORT);
        System.out.println("[server] listening on port " + PORT + "  (TCP)");
        System.out.println("[server] waiting for clients …\n");

        // Accept loop: one new thread per incoming connection
        while (true) {
            Socket socket = serverSocket.accept();   // blocks until a client connects
            int id = allocId();
            ClientHandler handler = new ClientHandler(socket, id);
            clients.put(id, handler);
            new Thread(handler).start();             // handle client on its own thread
        }
    }


    // ── per-client thread ─────────────────────────────────────────────────────

    /**
     * ClientHandler runs on its own thread and manages one client connection.
     *
     * Flow:
     *   1. Read the first line → treat it as the client's display name.
     *   2. Read subsequent lines → broadcast to all other clients.
     *   3. When the socket closes → unregister and notify others.
     */
    static class ClientHandler implements Runnable {

        final Socket socket;
        final int connId;
        final OutputStream out;       // kept for broadcasting back to this client
        String username = null;

        ClientHandler(Socket socket, int connId) throws IOException {
            this.socket = socket;
            this.connId = connId;
            this.out = socket.getOutputStream();
            System.out.println("[server] client #" + connId + " connected from "
                    + socket.getRemoteSocketAddress()
                    + "  (active=" + clients.size() + ")");
        }

        @Override
        public void run() {
            try (BufferedReader reader =
                         new BufferedReader(new InputStreamReader(socket.getInputStream()))) {

                String line;
                while ((line = reader.readLine()) != null) {   // readLine() blocks until '\n'
                    line = line.trim();
                    if (line.isEmpty()) continue;

                    if (username == null) {
                        // First message sets the display name
                        username = line;
                        System.out.println("[server] client #" + connId
                                + " → name '" + username + "'");
                        broadcast("*** " + username + " joined the chat ***", connId);
                    } else {
                        broadcast(username + ": " + line, connId);
                    }
                }

            } catch (IOException e) {
                // Client disconnected abruptly — normal, not an error
            } finally {
                clients.remove(connId);
                String name = (username != null) ? username : "#" + connId;
                System.out.println("[server] " + name
                        + " disconnected  (active=" + clients.size() + ")");
                broadcast("*** " + name + " left the chat ***", connId);
                try { socket.close(); } catch (IOException ignored) {}
            }
        }
    }
}

File: chat_example/tcp/python/server_tcp.py
"""
server_tcp.py — TCP Chat Server
Computer Networks Lab

Listens for TCP client connections on port 5555.
Every message received from one client is broadcast to all other clients.

Usage:
    python server_tcp.py
"""

import asyncio
from typing import Dict, Optional

# ── global registry: connection_id → writer ──────────────────────────────────
_writers: Dict[int, asyncio.StreamWriter] = {}
_usernames: Dict[int, str] = {}
_id_counter = 0

def _alloc_id() -> int:
    global _id_counter
    _id_counter += 1
    return _id_counter


# ── broadcast helper ──────────────────────────────────────────────────────────

def _broadcast(message: str, exclude: int):
    """Send *message* to every connected client except the one with id *exclude*."""
    data = (message + "\n").encode()
    for cid, writer in list(_writers.items()):
        if cid == exclude:
            continue
        try:
            writer.write(data)
        except Exception as e:
            print(f"[server] could not reach #{cid}: {e}")


# ── per-connection handler ────────────────────────────────────────────────────

async def handle_client(reader: asyncio.StreamReader, writer: asyncio.StreamWriter):
    """
    Called by asyncio for every new TCP connection.

    asyncio.start_server() calls this coroutine with a StreamReader/StreamWriter
    pair.  We loop reading lines until the client disconnects.
    """
    conn_id = _alloc_id()
    addr = writer.get_extra_info("peername")
    _writers[conn_id] = writer
    username: Optional[str] = None

    print(f"[server] client #{conn_id} connected from {addr}  "
          f"(active={len(_writers)})")

    try:
        while True:
            # Read one line at a time (messages end with '\n')
            data = await reader.readline()
            if not data:
                break                       # client closed the connection

            text = data.decode(errors="replace").strip()
            if not text:
                continue

            # First message from each client is their display name
            if username is None:
                username = text
                _usernames[conn_id] = username
                print(f"[server] client #{conn_id} → name '{username}'")
                _broadcast(f"*** {username} joined the chat ***", exclude=conn_id)
                continue

            line = f"{username}: {text}"
            print(f"[server] {line}")
            _broadcast(line, exclude=conn_id)

    except (asyncio.IncompleteReadError, ConnectionResetError):
        pass    # client disconnected abruptly

    finally:
        _writers.pop(conn_id, None)
        _usernames.pop(conn_id, None)
        name = username or f"#{conn_id}"
        print(f"[server] {name} disconnected  (active={len(_writers)})")
        _broadcast(f"*** {name} left the chat ***", exclude=conn_id)
        try:
            writer.close()
            await writer.wait_closed()
        except Exception:
            pass


# ── entry point ───────────────────────────────────────────────────────────────

async def main():
    HOST = "localhost"
    PORT = 5555

    # asyncio.start_server() opens a TCP socket and calls handle_client()
    # for every new incoming connection.
    server = await asyncio.start_server(handle_client, HOST, PORT)

    print(f"[server] listening on {HOST}:{PORT}  (TCP)")
    print("[server] waiting for clients …\n")

    async with server:
        await server.serve_forever()


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n[server] stopped.")
